modules/publicize/publicize.php

```diff
@@ -245,33 +407,146 @@ function show_options_popup( $service_name, $connection ) {
+			foreach ( $connections as $connection ) {
 
-		$connections = $this->get_connections( $service, $_blog_id, $_user_id );
-		return ( is_array( $connections ) && count( $connections ) > 0 ? true : false );
+				$id = $this->get_connection_id( $connection );
+
+				$connection_test_passed = true;
+				$connection_test_message = __( 'This connection is working correctly.' , 'jetpack' );
+				$user_can_refresh = false;
+				$refresh_text = '';
+				$refresh_url = '';
+
+				$connection_test_result = true;
+				if ( method_exists( $this, 'test_connection' ) ) {
+					$connection_test_result = $this->test_connection( $service_name, $connection );
+				}
+
+				if ( is_wp_error( $connection_test_result ) ) {
+					$connection_test_passed = false;
+					$connection_test_message = $connection_test_result->get_error_message();
+					$error_data = $connection_test_result->get_error_data();
+
+					$user_can_refresh = $error_data['user_can_refresh'];
+					$refresh_text = $error_data['refresh_text'];
+					$refresh_url = $error_data['refresh_url'];
+				}
+				// Mark facebook profiles as deprecated
+				if ( 'facebook' === $service_name ) {
+					if ( ! $this->is_valid_facebook_connection( $connection ) ) {
+						$connection_test_passed = false;
+						$user_can_refresh = false;
+						$connection_test_message = __( 'Facebook no longer supports Publicize connections to Facebook Profiles, but you can still connect Facebook Pages. Please select a Facebook Page to publish updates to.' );
+					}
+				}
+
+				$unique_id = null;
+				if ( ! empty( $connection->unique_id ) ) {
+					$unique_id = $connection->unique_id;
+				} else if ( ! empty( $connection['connection_data']['token_id'] ) ) {
+					$unique_id = $connection['connection_data']['token_id'];
+				}
+
+				$test_results[] = array(
+					'connectionID'          => $id,
+					'serviceName'           => $service_name,
+					'connectionTestPassed'  => $connection_test_passed,
+					'connectionTestMessage' => esc_attr( $connection_test_message ),
+					'userCanRefresh'        => $user_can_refresh,
+					'refreshText'           => esc_attr( $refresh_text ),
+					'refreshURL'            => $refresh_url,
+					'unique_id'             => $unique_id,
+				);
+			}
+		}
+
+		wp_send_json_success( $test_results );
 	}
 
 	/**
-	* Fires when a post is saved, checks conditions and saves state in postmeta so that it
-	* can be picked up later by @see ::publicize_post() on WordPress.com codebase.
-	*/
+	 * Run the connection test for the Connection
+	 *
+	 * @param string $service_name 'facebook', 'twitter', etc.
+	 * @param object|array The Connection object (WordPress.com) or array (Jetpack)
+	 * @return WP_Error|true WP_Error on failure. True on success
+	 */
+	abstract function test_connection( $service_name, $connection );
+
+/*
+ * Site Data
+ */
+
+	function user_id() {
+		return get_current_user_id();
+	}
+
+	function blog_id() {
+		return get_current_blog_id();
+	}
+
+/*
+ * Posts
+ */
```

Comment:
Purposefully indented like this?

Reply:
[Yes](https://github.com/Automattic/jetpack/pull/10043#discussion_r227519433).